### PR TITLE
(Feature #3059) Mail reports on a per user basis

### PR DESCRIPTION
**Pull request by dLobatog**
This pull requests adds a new rake task within the namespace reports that sends a digest of the status of the machines every user owns.

### CODE REVIEW

**Comment by domcleal** [comment]
[test]

**Comment by domcleal** [comment]
[test]

**Comment by ohadlevy** [comment]
@eLobato can you rebase please (after the 3.2 upgrade) thanks!

**Comment by dLobatog** [comment]
rebased

**Comment by dLobatog** [comment]
Test? I think it was properly rebased, I can rebase again against the latest if needed.

File: app/models/host_mailer.rb [hunk 1]
```diff
@@ -54,16 +45,54 @@ def summary(options = {})
...
+    content_type "text/html"
+    @hosts = host_data.sort_by { |h| h[1][:metrics]['failed'] }.reverse
+    @timerange = time
+    @out_of_sync = Host.out_of_sync.select { |h| h.owner == user }
+    @disabled = Host.alerts_disabled.select { |h| h.owner == user }
+  end
+
+  def set_url
+    if (@url = Setting[:foreman_url]).empty?
+      raise ":foreman_url: entry in Foreman configuration file, see http://theforeman.org/projects/foreman/wiki/Mail_Notifications"
```
Comment (GregSutcliffe) [inline app/models/host_mailer.rb]: This is in the UI now, not the config file, could be reworded - perhaps

```
:foreman_url is not set, please configure in the Foreman Web UI (More -> Settings -> General)
```

File: app/models/host_mailer.rb [hunk 2]
```diff
@@ -54,16 +45,54 @@ def summary(options = {})
     @out_of_sync = hosts.out_of_sync
     @disabled = hosts.alerts_disabled
     @filter = filter
-    mail(:to => email, :subnet => subject)
```
Comment (GregSutcliffe) [inline app/models/host_mailer.rb]: Why remove this?

**Comment by GregSutcliffe** [comment]
@eLobato Generally, looks good. Few small comments, and I think it might need another rebase. @domcleal do we need to apply i18n to rake tasks?

**Comment by domcleal** [comment]
@GregSutcliffe no i18n in rake tasks (yet)

**Comment by GregSutcliffe** [comment]
Cool, I'll try and test this tomorrow morning

File: lib/tasks/reports.rake [hunk 3]
```diff
@@ -78,3 +78,34 @@ namespace :reports do
...
+    options = {}
+
+    time = ENV['hours'].to_i.hours.ago if ENV['hours']
+    time = ENV['days'].to_i.days.ago if ENV['days']
+    options[:time] = time if time
+
+    users = User.select { |u| u.hosts.length > 0 }
+
+    users.each do |user|
+      HostMailer.deliver_failed_runs(user, options)
```
Comment (GregSutcliffe) [inline lib/tasks/reports.rake]: Typo here? The method add to HostMailer seems to be called `failed_runs` - I get a NoMethodError as it stands.

Comment (GregSutcliffe) [inline lib/tasks/reports.rake]: Also the method doesn't want two params, just one Hash.

**Comment by GregSutcliffe** [comment]
@eLobato couple of things are stopping me testing this, plus it needs a rebase. back to you :)

File: lib/tasks/reports.rake (hunk 3 above)
Comment (dLobatog) [inline lib/tasks/reports.rake]: I think rails was updated before I submitted this, as deliver_ methods don't work anymore. The rest of the report methods also make use of that (old) feature, I'll update this with a fix for that.

From [the rails guide](http://guides.rubyonrails.org/action_mailer_basics.html):

> In previous versions of Rails, you would call deliver_welcome_email or create_welcome_email. This has been deprecated in Rails 3.0 in favour of just calling the method name itself.

**Comment by dLobatog** [comment]
[test]

**Comment by dLobatog** [comment]
Test error is 'Could not find gem 'rails (= 3.2.13) ruby' in the gems available on this machine.' only on sqlite3, ruby 2.0.0, I haven't touched the Gemfile so maybe Jenkins is acting up?

**Comment by dLobatog** [comment]
[test]

**Comment by domcleal** [comment]
@eLobato it was a random network error while installing bundler.  Using [ test ] isn't necessary normally (Jenkins checks every 15 mins for new or updated PRs), but it's useful if you need to retest when that happens.

**Comment by dLobatog** [comment]
Rebased against develop

**Comment by ohadlevy** [comment]
@GregSutcliffe mind give this a final review? thanks!

**Comment by lzap** [comment]
And redmine ticket please, thank you.

LZ

On Sun, Sep 08, 2013 at 01:41:46AM -0700, Ohad Levy wrote:

> @GregSutcliffe mind give this a final review? thanks!
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/theforeman/foreman/pull/474#issuecomment-24017285

## 

Later,

 Lukas "lzap" Zapletal
 irc: lzap #theforeman

**Comment by GregSutcliffe** [comment]
@ohadlevy looks sane to me. as @lzap says can we get a RFE for this?

**Comment by dLobatog** [comment]
I just created a ticket on the tracker for this and rebased this branch against develop for a smooth merge.

**Comment by domcleal** [comment]
Thanks for the contribution @eLobato.  Merged as d70eecf for Foreman 1.3.

**Pull request closed without merging**